primitives/src/transaction.rs

```diff
@@ -0,0 +1,38 @@
+// SPDX-License-Identifier: CC0-1.0
+
+//! Bitcoin transactions.
+//!
+//! A transaction describes a transfer of money. It consumes previously-unspent
+//! transaction outputs and produces new ones, satisfying the condition to spend
+//! the old outputs (typically a digital signature with a specific key must be
+//! provided) and defining the condition to spend the new ones. The use of digital
+//! signatures ensures that coins cannot be spent by unauthorized parties.
+//!
+//! This module provides the structures and functions needed to support transactions.
+
+use core::fmt;
+
+/// The transaction version.
+///
+/// Currently, as specified by [BIP-68], only version 1 and 2 are considered standard.
+///
+/// Standardness of the inner `i32` is not an invariant because you are free to create transactions
+/// of any version, transactions with non-standard version numbers will not be relayed by the
+/// Bitcoin network.
+///
+/// [BIP-68]: https://github.com/bitcoin/bips/blob/master/bip-0068.mediawiki
+#[derive(Copy, PartialEq, Eq, Clone, Debug, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub struct Version(pub i32);
```

Comment:
FTR the existence of `non_standard` method suggests this field really shouldn't be `pub`.

Reply:
We have docs explaining why this is `pub`

```rust
/// Standardness of the inner `i32` is not an invariant because you are free to create transactions
/// of any version, transactions with non-standard version numbers will not be relayed by the
/// Bitcoin network.
```

Reply:
Its a thin wrapper that does not enforce invariants, I think the decision to make it public was correct. We should lead users to the right place but not force our opinions on them too heavily, right?

Reply:
I think Kix's point is that it's a little weird to have a public field and also a constructor which is equivalent to using the public field.

Personally I think it's fine. The constructor is more explicit about what's going on and the public field demonstrates that there are no invariants.

Reply:
That's not really my point. If you write `Version(42)` then it's in no way explicitly acknowledging that the version is non-standard, however we do have a constructor that does explicitly acknowledge it and we have constants to produce standard version. However the `non_standard` constructor is also misleading because someone might pass in 0 or 1, therefore it should be called `maybe_non_standard` if we want to have it.

Maybe it's just easier to remove the constructor though. I still like stronger API more but given that people tend to be pissed about these things, I wouldn't push it too hard in this case.

Reply:
Ah I see, those are all good points, being able to write `Version::non_standard(0)` seems to go against the point of having the constructor.